### PR TITLE
Hardcoded sorting of query results

### PR DESCRIPTION
**Pull request by GayathriVenkatesh**
Added sorting for instant query based on values and for range queries based on the ranking sum, both in descending order.

### CODE REVIEW

**Comment by codesome** [comment]
(For maintainers: target branch is not master, we are experimenting with post processing and these PRs pointing not to master are for testing the CI/practicing/etc. Hence you can ignore.)

File: web/api/v1/api.go [hunk 1]
```diff
@@ -372,6 +372,24 @@ func (api *API) query(r *http.Request) (result apiFuncResult) {
 		return apiFuncResult{nil, returnAPIError(res.Err), res.Warnings, qry.Close}
 	}
 
+	if res.Value.Type() == "vector" {
+		vec, ok := res.Value.(promql.Vector)
+		if !ok {
+			panic("Error converting to vector")
```
Comment (bwplotka) [inline web/api/v1/api.go:378]: Let's never use panics (:

File: web/api/v1/api.go [hunk 2]
```diff
@@ -372,6 +372,24 @@ func (api *API) query(r *http.Request) (result apiFuncResult) {
 		return apiFuncResult{nil, returnAPIError(res.Err), res.Warnings, qry.Close}
 	}
 
+	if res.Value.Type() == "vector" {
```
Comment (bwplotka) [inline web/api/v1/api.go:375]: It's usually better to use some `const` than hardcoded strings. Maybe we can pull it out from some package? (: (e.g `promql` or `model`

File: web/api/v1/api.go [hunk 3]
```diff
@@ -372,6 +372,24 @@ func (api *API) query(r *http.Request) (result apiFuncResult) {
...
+		sort.Slice(vec, func(i, j int) bool {
+			return vec[i].V > vec[j].V
+		})
+	} else if res.Value.Type() == "matrix" {
+		mat, ok := res.Value.(promql.Matrix)
+		if !ok {
+			panic("Error converting to matrix")
+		}
+		sort.Slice(mat, func(i, j int) bool {
+			return (sum(mat[i]) > sum(mat[j]))
```
Comment (bwplotka) [inline web/api/v1/api.go:389]: nice! I think let's go one step further and make sure the ranking function is under some variable / interface so we can quickly switch it  (: WDYT?

Maybe I want sum, maybe avg, maybe last sample, maybe first sample (:

File: web/api/v1/api.go [hunk 4]
```diff
@@ -452,6 +478,16 @@ func (api *API) queryRange(r *http.Request) (result apiFuncResult) {
 		return apiFuncResult{nil, returnAPIError(res.Err), res.Warnings, qry.Close}
 	}
 
+	if res.Value.Type() == "matrix" {
```
Comment (bwplotka) [inline web/api/v1/api.go:481]: ditto

File: web/api/v1/api_test.go [hunk 5]
```diff
@@ -613,6 +613,91 @@ func testEndpoints(t *testing.T, api *API, tr *testTargetRetriever, testLabelAPI
...
+						Metric: nil,
+					},
+				},
+			},
+		},
+		{
+			endpoint: api.queryRange,
+			query: url.Values{
+				"query": []string{"prometheus_target_interval_length_seconds"},
+				"start": []string{"0"},
```
Comment (bwplotka) [inline web/api/v1/api_test.go:667]: How we are sure that the mocked data is not giving us series in the right order before sorting? The quick way is to remove sort and check out the result (:

**Review by bwplotka — COMMENTED** [review]
Awesome! Good job, some suggestions only we made together with @codesome on our offline meeting.

File: web/api/v1/api.go (hunk 3 above)
Comment (GayathriVenkatesh) [reply web/api/v1/api.go]: Yes that makes sense! Will add that in :slightly_smiling_face:

File: web/api/v1/api.go [hunk 6]
```diff
@@ -195,6 +195,11 @@ type API struct {
 	gatherer                  prometheus.Gatherer
 }
 
+type RankingFunction interface {
+	sum(v promql.Series) float64
+	avg(v promql.Series) float64
+}
```
Comment (codesome) [inline web/api/v1/api.go:201]: Interface should show a generic structure and should not know about specific implementations. What about this?
```suggestion
type RankingFunction interface {
	rank(v promql.Series) float64
}
```

Additionally, we could just have the below and ranking would be just a function.
```
type RankingFunction func(promql.Series) float64
```

File: web/api/v1/api.go [hunk 7]
```diff
@@ -195,6 +195,11 @@ type API struct {
 	gatherer                  prometheus.Gatherer
 }
 
+type RankingFunction interface {
```
Comment (bwplotka) [inline web/api/v1/api.go:198]: Let's change this to something like 

```
type Ranker interface{
Rank(promql.Series) float64
}
```

as mentioned by @codesome  in our sync :hugs:

**Review by bwplotka — COMMENTED** [review]
LGTM, just interface nit (: Thanks good job :muscle:

File: web/api/v1/api.go [hunk 8]
```diff
@@ -333,6 +337,9 @@ func (api *API) options(r *http.Request) apiFuncResult {
 	return apiFuncResult{nil, nil, nil, nil}
 }
 
+type sum struct{}
+type avg struct{}
+
```
Comment (codesome) [inline web/api/v1/api.go:342]: We keep the declarations of structs with their methods usually, so you can have `sum` just before its implementation and `avg` similarly (in Go you don't need the type to be declared before the usage, you can have it anywhere).

File: web/api/v1/api.go [hunk 9]
```diff
@@ -371,6 +378,27 @@ func (api *API) query(r *http.Request) (result apiFuncResult) {
 	if res.Err != nil {
 		return apiFuncResult{nil, returnAPIError(res.Err), res.Warnings, qry.Close}
 	}
+	var ranker Ranker = sum{}
+
+	if res.Value.Type() == parser.ValueTypeVector {
+		vec, ok := res.Value.(promql.Vector)
+		if !ok {
+			err := errors.New("Cannot convert to vector")
```
Comment (codesome) [inline web/api/v1/api.go:386]: Nit: We start error message with a smaller case letter but logs with upper case letters, unless for acronyms :) (same below)

**Review by codesome — APPROVED** [review]
LGTM! Some nits and lets merge this!

File: web/api/v1/api.go [hunk 10]
```diff
@@ -371,6 +378,27 @@ func (api *API) query(r *http.Request) (result apiFuncResult) {
 	if res.Err != nil {
 		return apiFuncResult{nil, returnAPIError(res.Err), res.Warnings, qry.Close}
 	}
+	var ranker Ranker = sum{}
```
Comment (codesome) [inline web/api/v1/api.go:381]: Lets use `avg{}` here to avoid test failure (since it uses `sum{}` internally)

**Pull request merged**